src/main/java/com/imgix/URLBuilder.java

```diff
@@ -73,72 +79,252 @@ public String createURL(String path, Map<String, String> params) {
+        StringBuilder srcset = new StringBuilder();
 
-        for (int ratio: srcsetTargetRatios) {
+        for (int ratio: TARGET_RATIOS) {
             params.put("dpr", Integer.toString(ratio));
-            srcset += this.createURL(path, params) + " " + ratio + "x,\n";
+
+            if (!disableVariableQuality) {
+                params.put("q", DPR_QUALITIES[ratio].toString());
+            }
+            srcset.append(this.createURL(path, params)).append(" ").append(ratio).append("x,\n");
         }
 
         return srcset.substring(0, srcset.length() - 2);
     }
 
-    private static ArrayList<Integer> targetWidths() {
+    /**
+     * Create an `ArrayList` of integer target widths.
+     *
+     * If `begin`, `end`, and `tol` are the default values, then the
+     * targets are not custom, in which case the default widths are
+     * returned.
+     *
+     * A target width list of length one is valid: if `begin` == `end`
+     * then the list begins where it ends.
+     *
+     * When the `while` loop terminates, `begin` is greater than `end`
+     * (where `end` less than or equal to `MAX_WIDTH`). This means that
+     * the most recently appended value may, or may not, be the `end`
+     * value.
+     *
+     * To be inclusive of the ending value, we check for this case and the
+     * value is added if necessary.
+     *
+     * @param begin - beginning image width value
+     * @param end - ending image width value
+     * @param tol - tolerable amount of width value variation
+     * @return array list of image width values
+     */
+    public static ArrayList<Integer> targetWidths(int begin, int end, int tol) {
+        return computeTargetWidths(begin, end, tol);
+    }
+
+    public static ArrayList<Integer> targetWidths() {
+        double begin = MIN_WIDTH, end = MAX_WIDTH;
+
+        ArrayList<Integer> resolutions = new ArrayList<Integer>();
+        while (begin < end) {
+            resolutions.add((int) Math.round(begin));
+            begin *= 1 + ((double) SRCSET_WIDTH_TOLERANCE / 100) * 2;
+        }
+
+        int lastIndex = resolutions.size() - 1;
+        if (resolutions.get(lastIndex) < end) {
+            resolutions.add((int) end);
+        }
+
+        return resolutions;
+    }
+
+    /**
+     * Create an `ArrayList` of integer target widths.
+     *
+     * This function is the implementation details of `targetWidths`.
+     * This function exists to provide a consistent interface for
+     * callers of `targetWidths`.
+     *
+     * This function implements the syntax that fulfills the semantics
+     * of `targetWidths`. Meaning, `begin`, `end`, and `tol` are
+     * to be whole integers, but computation requires `double`s. This
+     * function hides this detail from callers.
+     */
+    private static ArrayList<Integer> computeTargetWidths(double begin, double end, double tol) {
+        if (notCustom(begin, end, tol)) {
+            return targetWidths();
```

Comment:
could this be replaced with `SRCSET_TARGET_WIDTHS`?

Reply:
Yeah, I think that would be ideal, but `SRCSET_TARGET_WIDTHS` is currently being generated via the same function, `targetWidths()`. This is a hold over from before this PR. Would you be up for making this constant an array?

Reply:
Yeah my gut feeling says that we should.

Reply:
Cool, yeah, I've been bringing it up a lot, but haven't gotten a solid from anyone––this seals it!